shimmingtoolbox/coils/spherical_harmonics.py

```diff
@@ -0,0 +1,136 @@
+
+        0th-to-2nd order terms inclusive
+
+        >>> orders = np.array(range(0, 3))
+        >>> basis = spherical_harmonics(orders, x, y, z)
+
+    Notes:
+        - basis[:, :, :,0] corresponds to the 0th-order constant term (globally=unity)
+            - 0: *c*
+
+        - basis[:, :, :, 1:3] to 1st-order linear terms
+            - 1: *y*
+            - 2: *z*
+            - 3: *x*
+
+        - basis[:, :, :, 4:8] to 2nd-order terms
+            - 4: *xy*
+            - 5: *zy*
+            - 6: *z2*
+            - 7: *zx*
+            - 8: *x2y2*
+
+        Based on
+            - spherical_harmonics.m by topfer@ualberta.ca
+            - calc_spherical_harmonics_arb_points_cz.m by jaystock@nmr.mgh.harvard.edu
+    """
+
+    def number_of_coef(orders):
+        out = 0
+        for n in orders:
+            m = 2 * n + 1
+            out += m
+
+        return out
+
+    def leg_rec_harmonic_cz(n, m, pos_x, pos_y, pos_z):
+        """
+        Returns harmonic field for the required solid harmonic addressed by n, m based on the Legendre
+        polynomial calculation. Positive m values correspond to the cosine component, negative to the sine. Returned
+        fields will eventually follow RRI s convention
+        """
+
+        r2 = pos_x ** 2 + pos_y ** 2 + pos_z ** 2
+        r = np.sqrt(r2)
+        phi = np.arctan2(pos_y, pos_x)
+        cos_theta = np.cos(np.arctan2(np.sqrt(pos_x ** 2 + pos_y ** 2), pos_z))
+        # cos_theta = pos_z / r
+
+        if m >= 0:
+            c = 1
+        else:
+            c = 0
+            m = -m
+
+        # y_mn = leg_rec(n, m, cos_theta)  # Does the same as lpmv function (Will remove after more tests)
+        y_mn = lpmv(m, n, cos_theta)
+
+        rri_norm = math.factorial(n + m + 1) / math.factorial(n - m) / factorial2(2 * m)
+
+        out = (n + m + 1) * (r ** n) * (np.cos(m * phi) * c + np.sin(m * phi) * (1 - c)) * y_mn / rri_norm
+
+        return out
+
+    # Check inputs
+    if not (x.shape == y.shape == z.shape):
+        raise RuntimeError('Input arrays X, Y, and Z must be identically sized')
+
+    if x.ndim == 3:
+        grid_size = x.shape
+    else:
+        raise RuntimeError('Input arrays X, Y, and Z must have 3 dimensions')
+
+    if not np.all(orders >= 0):
+        raise RuntimeError('Orders must be positive')
+
+    # Initialize variables
+    n_voxels = x.size
+    n_orders = orders.size
+    n_basis = number_of_coef(orders)
+    harm_all = np.zeros([n_voxels, n_basis])
+
+    ii = 0
+
+    # Compute basis
+    for iOrder in range(0, n_orders):
+
+        n = orders[iOrder]
+        if n == 0:
+            m = np.zeros([1], dtype=int)
+        else:
+            m = np.array(range(-orders[iOrder], orders[iOrder] + 1, 1))
+
+        for mm in range(0, m.size):
+
+            # The first 2 * n(1) + 1 columns of the output correspond to harmonics of order n(1), and the next
+            # 2 * n(2) + 1 columns correspond to harmonics of order n(2), etc.
```

Comment:
These are good clarifying comments, especially for contributors like me that have no background in MRI physics 🙂